### PR TITLE
Fixed deleting monitor workflow metadata

### PR DESCRIPTION
**Pull request by stevanbz**
*Issue #, if available:*
Deleting the workflow and potential delete of it's metadata.

*Description of changes:*
This PR handles workflow removal, removal of it's delegate monitors - deleting the monitors, theirs metadata, potentially queries from percolate index (if the delegate is doc level monitor), and removal of workflow metadata.

Workflow execution will cause underlying monitor metadata to be generated. Beside deleting the workflow, user can delete the delegate monitors and their metadata - if the delegate monitor is being deleted, all it's metadata should be deleted: metadata related with the given workflow execution and metadata related with the monitor execution.

If user has cluster privilege for deleting the workflows he should be able to delete the delegate monitors that are part of the given workflow. 
Once the workflow deletion starts, the context is being stashed meaning that the program in the section when deleting the workflows, monitors and metadata won't have any thread context related data. Namely, when starting alerting transport actions 
next check is being called: if (!validateUserBackendRoles(user, actionListener)) {
return
}
This check first checks if the the filterBy is enabled and then check if the user is null -- if both of them are true - the error is raised.
The check prevents to call one transport action from another with stashed context - since it will raise an error (because when the context is stashed the user will be null). In order to avoid this situation, the common parts of the code related with checking if the monitor is part of some workflow, deleting the monitors, their metadata and queries is extracted to a helper/utility class named: [MonitorDeleteService](https://github.com/opensearch-project/alerting/pull/882/files#diff-76485686809f4b34c3520d27d44ad9f57ae66f2bb3875577906f0906820cd1dcR44).

*CheckList:*
[ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/model/MonitorMetadata.kt [hunk 1]
```diff
@@ -120,9 +120,10 @@ data class MonitorMetadata(
             return MonitorMetadata(sin)
         }
 
-        fun getId(monitor: Monitor, workflowId: String? = null): String {
-            return if (workflowId.isNullOrEmpty()) "${monitor.id}-metadata"
-            else "${monitor.id}-$workflowId-metadata"
+        fun getId(monitor: Monitor, workflowMetadataId: String? = null): String {
+            return if (workflowMetadataId.isNullOrEmpty()) "${monitor.id}-metadata"
+            // WorkflowMetadataId already contains -metadata suffix
+            else "${monitor.id}-$workflowMetadataId"
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/model/MonitorMetadata.kt:126]: what is the motivation behind this change

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt [hunk 2]
```diff
@@ -76,18 +76,12 @@ interface SecureTransportAction {
     }
 
     fun <T : Any> validateUserBackendRoles(user: User?, actionListener: ActionListener<T>): Boolean {
+        if (!filterByEnabled || user == null) {
+            return true
+        }
+
         if (filterByEnabled) {
```
Comment (stevanbz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt:83]: I think we maybe have an issue on alerting when starting the transport action when security is enabled and filterBy also.
Just wanted to check with you what's your opinion.

Namely, when starting alerting transport actions (like for example TransportIndexMonitorAction: https://github.com/opensearch-project/alerting/blob/main/alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportIndexMonitorAction.kt#L126-L128)
we are calling this check: if (!validateUserBackendRoles(user, actionListener)) {
return
}

This check first checks if the the filterBy is enabled and then check if the user is null -- if both of them are true - the error is raised.
Check it out here: https://github.com/opensearch-project/alerting/blob/main/alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt#L79-L82

On other plugins, for example on notification, this check is slightly different:
https://github.com/opensearch-project/notifications/blob/main/notifications/notifications/src/main/kotlin/org/opensearch/notifications/security/UserAccessManager.kt#L55-L57
or for example on index-management:
https://github.com/opensearch-project/index-management/blob/main/src/main/kotlin/org/opensearch/indexmanagement/util/SecurityUtils.kt#L137-L139

ie. it checks if the user is null OR if the filterBy is enabled and returns true.
The check we have on alerting will prevent us, to call one transport action from another with stashed context - since it will raise an error (because when the context is stashed the user will be null)

Should we adjust the check on alerting to reflect the same logic or? Tnx in advance!
@eirsep @lezzago

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 3]
```diff
@@ -150,26 +154,38 @@ class TransportDeleteWorkflowAction @Inject constructor(
...
+                    var deleteWorkflowResponse = DeleteWorkflowResponse(deleteResponse.id, deleteResponse.version)
+
+                    val workflowMetadataId = WorkflowMetadata.getId(workflow.id)
+
+                    val metadataIdsToDelete = mutableListOf(workflowMetadataId)
+
                     if (deleteDelegateMonitors == true) {
-                        if (user != null && filterByEnabled) {
-                            // Un-stash the context
-                            withClosableContext(
```
Comment (stevanbz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:157]: @eirsep @lezzago - if we conclude that we need to change the part I described in the comment above, we should definitely remove the context unstash so the program in next couple of lines, once we are deleting the monitor, can behave as an admin.

If we are going to keep the check in the comment above as it is, then the user must have monitor_delete cluster privilege - meaning that even before starting workflow deletion we would need to check it's cluster permission (which I think we shouldn't do and I don't know how to do since the security plugin does this I guess).

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt [hunk 4]
```diff
@@ -76,18 +76,12 @@ interface SecureTransportAction {
     }
 
     fun <T : Any> validateUserBackendRoles(user: User?, actionListener: ActionListener<T>): Boolean {
+        if (!filterByEnabled || user == null) {
```
Comment (stevanbz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt:79]: I think we maybe have an issue on alerting when starting the transport action when security is enabled and filterBy also.
Just wanted to check with you what's your opinion.

Namely, when starting alerting transport actions (like for example TransportIndexMonitorAction: https://github.com/opensearch-project/alerting/blob/main/alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportIndexMonitorAction.kt#L126-L128)
we are calling this check: if (!validateUserBackendRoles(user, actionListener)) {
return
}

This check first checks if the the filterBy is enabled and then check if the user is null -- if both of them are true - the error is raised.
Check it out here: https://github.com/opensearch-project/alerting/blob/main/alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt#L79-L82

On other plugins, for example on notification, this check is slightly different:
https://github.com/opensearch-project/notifications/blob/main/notifications/notifications/src/main/kotlin/org/opensearch/notifications/security/UserAccessManager.kt#L55-L57
or for example on index-management:
https://github.com/opensearch-project/index-management/blob/main/src/main/kotlin/org/opensearch/indexmanagement/util/SecurityUtils.kt#L137-L139

ie. it checks if the user is null OR if the filterBy is enabled and returns true.
The check we have on alerting will prevent us, to call one transport action from another with stashed context - since it will raise an error (because when the context is stashed the user will be null)

Should we adjust the check on alerting to reflect the same logic or? Tnx in advance!
@eirsep @lezzago

Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/SecureTransportAction.kt:79]: @getsaurabh02 Adding Saurabh so he can also participate

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt (hunk 3 above)
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:157]: @getsaurabh02 also, here

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 5]
```diff
@@ -150,26 +154,38 @@ class TransportDeleteWorkflowAction @Inject constructor(
...
-                            deleteMonitors(delegateMonitorIds, RefreshPolicy.IMMEDIATE)
-                        }
+                        val failedMonitorIds = tryDeletingMonitors(delegateMonitorIds, RefreshPolicy.IMMEDIATE)
+                        // Update delete workflow response
+                        deleteWorkflowResponse.nonDeletedMonitors = failedMonitorIds
+                        // Delete monitors workflow metadata
+                        // Monitor metadata will be in monitorId-workflowId-metadata format
+                        metadataIdsToDelete.addAll(deletableMonitors.map { MonitorMetadata.getId(it, workflowMetadataId) })
+                    }
+                    // Delete the monitors and it's metadata
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:172]: Update comment to say `// Delete the monitors' workflow metadata` as that is what it is actually doing.

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 6]
```diff
@@ -150,26 +154,38 @@ class TransportDeleteWorkflowAction @Inject constructor(
...
+                            .source(ScheduledJob.SCHEDULED_JOBS_INDEX)
+                            .filter(QueryBuilders.idsQuery().addIds(*metadataIdsToDelete.toTypedArray()))
+                            .execute(it)
                     }
-                    actionListener.onResponse(DeleteWorkflowResponse(deleteResponse.id, deleteResponse.version))
+
+                    if (deleteMonitorWorkflowMetadataResponse.isTimedOut) {
+                        actionListener.onFailure(
+                            AlertingException.wrap(
+                                OpenSearchException("Cannot determine that the ${ScheduledJob.SCHEDULED_JOBS_INDEX} index is healthy")
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:183]: Change the exception message. It could be something like: `Cannot delete monitors' workflow metadata due to timeout.`

**Review by lezzago — CHANGES_REQUESTED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/model/MonitorMetadata.kt [hunk 7]
```diff
@@ -120,9 +120,16 @@ data class MonitorMetadata(
...
 
-        fun getId(monitor: Monitor, workflowId: String? = null): String {
-            return if (workflowId.isNullOrEmpty()) "${monitor.id}-metadata"
-            else "${monitor.id}-$workflowId-metadata"
+        // workflowMetadataId is used as key for monitor metadata in the case when the workflow execution happens
+        // so the monitor lastRunContext (in the case of doc level monitor) is not interfering with the monitor execution
+        // WorkflowMetadataId will be either workflowId-metadata (when executing the workflow as it is scheduled)
+        // or timestampWithUUID-metadata (when the workflow is temp - ie. executed in a dry-run mode)
+        // In the case of temp workflow, doc level monitors must have lastRunContext created from scratch
+        // That's why we are using workflowMetadataId - in order to ensure that the doc level monitor metadata is created from scratch
```
[the remark this answers is not in the document]
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/model/MonitorMetadata.kt:128]: Got it. Tnx. Will do so

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt (hunk 3 above)
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:157]: @lezzago @eirsep @getsaurabh02 
Answered in previous response

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 8]
```diff
@@ -150,26 +154,38 @@ class TransportDeleteWorkflowAction @Inject constructor(
...
+                        val failedMonitorIds = tryDeletingMonitors(delegateMonitorIds, RefreshPolicy.IMMEDIATE)
+                        // Update delete workflow response
+                        deleteWorkflowResponse.nonDeletedMonitors = failedMonitorIds
+                        // Delete monitors workflow metadata
+                        // Monitor metadata will be in monitorId-workflowId-metadata format
+                        metadataIdsToDelete.addAll(deletableMonitors.map { MonitorMetadata.getId(it, workflowMetadataId) })
+                    }
+                    // Delete the monitors and it's metadata
+                    val deleteMonitorWorkflowMetadataResponse: BulkByScrollResponse = client.suspendUntil {
+                        DeleteByQueryRequestBuilder(client, DeleteByQueryAction.INSTANCE)
```
Comment (stevanbz) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:174]: Here should say:
 DeleteByQueryRequestBuilder(this, DeleteByQueryAction.INSTANCE)

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt (hunk 6 above)
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:183]: Makes sense

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 9]
```diff
@@ -193,24 +206,36 @@ class TransportDeleteWorkflowAction @Inject constructor(
...
-                val deleteRequest = DeleteMonitorRequest(monitorId, refreshPolicy)
-                val searchResponse: DeleteMonitorResponse = client.suspendUntil {
-                    AlertingPluginInterface.deleteMonitor(this as NodeClient, deleteRequest, it)
+            for (monitor in monitors) {
+                val monitorId = monitor.id
+                try {
+                    val monitorIsWorkflowDelegate = DeleteMonitorService.monitorIsWorkflowDelegate(monitorId)
+                    if (monitorIsWorkflowDelegate) {
+                        nonDeletedMonitorIds.add(monitorId)
+                    }
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:224]: Why is this needed? We already check if this monitor is not part of any other workflow.

**Review by lezzago — CHANGES_REQUESTED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/DeleteMonitorService.kt [hunk 10]
```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.alerting
```
[the remark this answers is not in the document]
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/DeleteMonitorService.kt:6]: Maybe if we create a new package named service?

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt (hunk 9 above)
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:224]: You are right. 
We already filtered the monitors that can be deleted(we can delete only the monitors belonging to a given workflow - if the monitor is shared between multiple workflows - those ones will be skipped)

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 11]
```diff
@@ -270,7 +264,7 @@ class TransportDeleteWorkflowAction @Inject constructor(
                 }
                 workflow.copy(id = hit.id, version = hit.version)
             }
-            val workflowMonitors = workflows.filter { it.id != workflowIdToBeDeleted }.flatMap { (it.inputs[0] as CompositeInput).getMonitorIds() }.distinct()
+            val workflowMonitors = workflows.flatMap { (it.inputs[0] as CompositeInput).getMonitorIds() }.distinct()
```
Comment (lezzago) [inline alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:267]: Why are we removing this filter? We need to ignore the workflow we want to delete.

File: alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt [hunk 12]
```diff
@@ -595,9 +648,31 @@ class WorkflowMonitorIT : WorkflowSingleNodeTestCase() {
...
+            e.message?.let {
+                assertTrue(
+                    "Exception not returning GetMonitor Action error ",
+                    it.contains("List is empty")
+                )
+            }
+        }
+
+        try {
+            searchMonitorMetadata(monitorMetadataId2)
```
Comment (lezzago) [inline alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt:664]: We need to validate the result to make sure it is deleted.

File: alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt [hunk 13]
```diff
@@ -595,9 +648,31 @@ class WorkflowMonitorIT : WorkflowSingleNodeTestCase() {
                 )
             }
         }
+        // Verify that the monitors metadata are deleted
+        try {
+            searchMonitorMetadata(monitorMetadataId1)
```
[the remark this answers is not in the document]
Comment (lezzago) [reply alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt:653]: I think the same issue is there with the usages of `searchWorkflowMetadata` and `getWorkflowById` for checking if it was deleted.

[the remark this answers is not in the document]
Comment (lezzago) [reply alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt:653]: If it doesnt go through the exception, does it mean it should fail the test? If so, we should add a `fail` right after `searchMonitorMetadata(monitorMetadataId1)`. Same for the similar comment below.

File: alerting/src/main/kotlin/org/opensearch/alerting/service/DeleteMonitorService.kt [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+/**
+ * Component used when deleting the monitors
+ */
+object DeleteMonitorService :
+    CoroutineScope by CoroutineScope(SupervisorJob() + Dispatchers.Default + CoroutineName("WorkflowMetadataService")) {
+    private val log = LogManager.getLogger(this.javaClass)
+
+    private lateinit var client: Client
+
+    fun initialize(
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/service/DeleteMonitorService.kt:51]: we can just accept client in the method argument and make this a DeleteMonitorUtils class.

Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/service/DeleteMonitorService.kt:51]: i will raise a PR making that change.

**Review by eirsep — CHANGES_REQUESTED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/service/DeleteMonitorService.kt (hunk 14 above)
Comment (stevanbz) [reply alerting/src/main/kotlin/org/opensearch/alerting/service/DeleteMonitorService.kt:51]: We can - makes sense to do this change. 
I just followed the pattern we used in MetadataService -> great. Tnx for making this change.

File: alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (eirsep) [reply alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt:653]: adding fail() check within try() as catch block asserts result list is empty

File: alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt (hunk 12 above)
Comment (eirsep) [reply alerting/src/test/kotlin/org/opensearch/alerting/WorkflowMonitorIT.kt:666]: adding fail() check within try() as catch block asserts result list is empty
 addressed https://github.com/opensearch-project/alerting/commit/a3105171b11147bc57241d1bf164477ae13e7fb4

File: alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt [hunk 15]
```diff
@@ -150,26 +151,38 @@ class TransportDeleteWorkflowAction @Inject constructor(
...
-                    actionListener.onResponse(DeleteWorkflowResponse(deleteResponse.id, deleteResponse.version))
+
+                    if (deleteMonitorWorkflowMetadataResponse.isTimedOut) {
+                        actionListener.onFailure(
+                            AlertingException.wrap(
+                                OpenSearchException("Cannot delete monitors workflow metadata due to timeout.")
+                            )
+                        )
+                        return
+                    }
```
[the remark this answers is not in the document]
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt:183]: Checking.. will handle holistically

[the remark this answers is not in the document]
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDeleteWorkflowAction.kt]: removed that check [6700a1e](https://github.com/opensearch-project/alerting/pull/882/commits/6700a1e241147afaecf0dd692d16a3d86c6cb054)

Agreed that timeout requires no special attention

**Review by lezzago — APPROVED** [review]
